---
title: "Lost samples (Freeze 10)"
author: "Wonji Kim (rewki@channing.harvard.edu)"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 6
    number_sections: true
---


***
# Settings

```{r setup, include=TRUE}
## Set R options
library(knitr)
library(DT)
library(dplyr)
library(data.table)

knitr::opts_chunk$set(echo = TRUE)
## knitr::opts_chunk$set(root.dir='/udd/rewki/projects/TOPMed/Freeze.10/Lost_samples')
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=TRUE)
options(stringsAsFactors=FALSE, digits=4)
```
- Working directory : ``r getwd()`` (It is linked to `/udd/rewki/projects/TOPMed/Freeze.10/Lost_samples`)


# Compiling samples for PLCOPD, EOCOPD and GLAXO
- List of samples to be dropped from freeze.10: samples whose keep_drop=="drop" in ANNOTATED_MANIFEST.csv
- List of samples suggested for re-assignment to a new subject
- List of samples from previous genotyping that are now "suspect" and considered-for-removal 



```{r}
get.list <- function(study=c("PLCOPD","EOCOPD","GLAXO")){
  library(data.table)
  annot.sam <- fread(paste0("/proj/regeps/regep00/studies/TopMed/data/dna/whole_genome/TopMed/data/freezes/freeze.10.cdnm/multiomics/",study,"/ANNOTATED_MANIFEST.csv"),head=T,data.table=F)
  #table(annot.sam$keep_drop,annot.sam$status)
  
  # cannot find following files..
  #sexcheck <- fread(paste0("tmp/",study,"/",study,"_annotated_plink_merged.sexcheck"),head=T,data.table=F)
  #het <- fread(paste0("tmp/",study,"/",study,"_annotated_plink_merged.het"),head=T,data.table=F)
  #imiss <- fread(paste0("tmp/",study,"/",study,"_annotated_plink_merged.imiss"),head=T,data.table=F)
  
  df <- annot.sam[annot.sam$keep_drop=="drop"|annot.sam$status=="review_unmatched_sample",]
  df$study <- study
  return(df)
}

dfs <- do.call(rbind,lapply(c("PLCOPD","EOCOPD","GLAXO"),get.list))

# Samples to be dropped
drops <- dfs[dfs$keep_drop=="drop",]
dim(drops)

# Samples unmatched to the assigned study
unmatched <- dfs[dfs$status=="review_unmatched_sample",]
dim(unmatched)

write.table(drops,"drop_f10.txt",row.names=F,quote=F)
write.table(unmatched,"unmatched_f10.txt",row.names=F,quote=F)
```


# 83 lost samples

## Cross study check
- 83 samples vs other TOPMed cohorts (GECOPD, ECLIPSE, EOCOPD, GLAXO, LTRC and PLCOPD)
- Workflow
    1. Extract pruned biallelic SNPs with MAF > 5% of 83 lost sampels and each study from TOPMed freeze10a.IRC
    2. Run KING. 

### Extract study samples + 83 samples
```{r}
library(data.table)
  
losts <- fread("/udd/rewki/projects/TOPMed/Freeze.10/Lost_samples/copd_unknown_nwdids_83.txt",head=F,data.table=F)

f10a.irc <- fread("/proj/regeps/regep00/studies/TopMed/data/dna/whole_genome/TopMed/data/freezes/freeze.10a.irc/manifests/nwdids.txt",head=F,data.table=F)
f10a.irc.copd <- fread("/proj/regeps/regep00/studies/TopMed/data/dna/whole_genome/TopMed/data/freezes/freeze.10a.irc/manifests/stashq.txt",head=F,data.table=F)

dim(f10a.irc)
dim(f10a.irc.copd)  # 286 samples were not included here
sum(f10a.irc$V1%in%losts$V1)
sum(f10a.irc.copd$V1%in%losts$V1)

STUDY <- unique(f10a.irc.copd$V5)
for(study in STUDY){
  all.sp <- c(f10a.irc.copd$V1[f10a.irc.copd$V5==study],losts$V1)
  if(study=="ECLPSE") study <- "ECLIPSE"
  write.table(all.sp,paste0(study,"_samples.txt"),row.names=F,col.names=F,quote=F)
}
```

```

# Extract biallelic common SNPs..
sh extract_samples.sh

# Convert bcf files to pgen files
sh convert_pgen.sh

# Merge files 
sh merge_plink.sh
```



## Concordance check using KING
```
sh King.sh
```

```{r echo=F}
library(data.table)
losts <- fread("/udd/rewki/projects/TOPMed/Freeze.10/Lost_samples/copd_unknown_nwdids_83.txt",head=F,data.table=F)
losts$new.study <- ""



for(study in c("GECOPD","EOCOPD","ECLIPSE","GLAXO","PLCOPD","LTRC")){
  cat(paste0("#################### ",study," ####################\n"))

  if(file.exists(paste0("king_",study,".con"))){
    concord <- read.table(paste0("king_",study,".con"),head=T) 
    concord.losts <- concord[concord$ID1%in%losts$V1 | concord$ID1%in%losts$V2,]
    cat(paste0("Among ",nrow(concord)," pairs of duplicates, ",nrow(concord.losts)," pairs contain *Lost* samples.\n"))
    if(nrow(concord.losts)>0){
      print(kable(concord.losts))
      cat("\n\n")
      losts$new.study[losts$V1%in%concord.losts$ID1|losts$V1%in%concord.losts$ID2] <- paste0(losts$new.study[losts$V1%in%concord.losts$ID1|losts$V1%in%concord.losts$ID2],";",study)
    } else {
      cat("\n")
    }
  } else {
    cat("No duplicates are found with heterozygote concordance rate > 80%.\n\n")
  }
}

losts$new.study <- gsub("^;","",losts$new.study)
cat(paste0(sum(losts$new.study!="")," subjects have duplicates in the COPD studies.\n"))
print(kable(losts[losts$new.study!="",]))

losts$source <- ""
losts$source[losts$new.study!=""] <- "Cross_study_check"
colnames(losts)[1] <- "NWD_ID"
write.csv(losts,"cross.New_study_83samples.csv",row.names=F,quote=F)
```

- NWD912732(GECOPD) - NWD790941(ECLIPSE)

## Independent external reference dataset
- Workflow
    1. Extract pruned biallelic SNPs with MAC > 5% of 83 samples from TOPMed freeze 10
    2. Liftover independent reference dataset from hg19 to hg38 if needed
    3. Extract overlapping SNVs between 83 samples and independent reference dataset
    4. Merge datasets
    5. Run KING

- Reference dataset
    - COPDGene genotyping data
        - /proj/regeps/regep00/studies/COPDGene/data/gwas/CG10k_[AA,NHW]_hg19_Oct2017.[bed,bim,fam]
    - COPDGene exome chip data
        - /proj/regeps/regep00/studies/COPDGene/data/exome/CleanExomeChip/data/canonical/[AA,NHW]exomeQCed.[bed,bim,fam]
    - EOCOPD, ICGN, TCGS
        - /proj/regeps/regep00/studies/EOCOPD/data/dna/exome_chip/BWH_Silverman_Exome6k/data/canonical/
        - exome6kSubAndMarkCleanV03.bed
        - exome6kSubAndMarkCleanV03_renamed.bim
        - exome6kSubAndMarkCleanV03.fam
    - LTRC
        - /proj/regeps/regep00/studies/LTRC/analyses/resaf/Genotypes/Freeze9b_redo/PLINK

### Extract 83 samples from TOPMed WGS data
```

# Extract biallelic common SNPs..
sh extract_samples.sh

# Merge files
sh merge_plink.sh
```


### COPDGene genotyping data 
- hg38: /proj/edith/regeps/regep00/studies/COPDGene/analyses/rewki/Concordance/1.GWAS/1.Overlapping/1.liftover/GWAS_[NHW,AA]_hg38.[bed,bim,fam]

```{r eval=F}
do.king <- function(race=c("AA","NHW")){
  library(data.table)
  copdgene <- fread(paste0("/proj/edith/regeps/regep00/studies/COPDGene/analyses/rewki/Concordance/1.GWAS/1.Overlapping/1.liftover/GWAS_",race,"_hg38.bim"),head=F,data.table=F)
  lost_sp <- fread(paste0("/udd/rewki/projects/TOPMed/Freeze.10/Lost_samples/2.indep_study_check/copd_unknown_83_merge_pruned.bim"),head=F,data.table=F)
  
  copdgene$snpID <- with(copdgene,paste(V1,V4,V5,V6,sep=":"))
  lost_sp$snpID1 <- with(lost_sp,paste(V1,V4,V5,V6,sep=":")) 
  lost_sp$snpID2 <- with(lost_sp,paste(V1,V4,V6,V5,sep=":")) 

  ovl <- copdgene[copdgene$snpID%in%lost_sp$snpID1 | copdgene$snpID%in%lost_sp$snpID2,]
  ovl.bed <- ovl[,c("V1","V4","V4")]

  write.table(ovl.bed,paste0(race,"_plink.bed"),row.names=F,col.names=F,quote=F)
  
  ### Extract overlapping variants
  ## COPDGene
  system(paste0("plink2 --bfile /proj/edith/regeps/regep00/studies/COPDGene/analyses/rewki/Concordance/1.GWAS/1.Overlapping/1.liftover/GWAS_",race,"_hg38 --make-pgen --sort-vars --out GWAS_",race,"_hg38"))
  system(paste0("plink2 --pfile GWAS_",race,"_hg38 --extract bed1 ",race,"_plink.bed --make-bed --out COPDGene_",race))
  bim <- fread(paste0("COPDGene_",race,".bim"),head=F,data.table=F)
  bim$V2 <- with(bim,paste(V1,V4,sep=":"))
  write.table(bim,paste0("COPDGene_",race,".bim"),row.names=F,col.names=F,quote=F,sep="\t")

  ## 83 samples
  system(paste0("plink2 --bfile /udd/rewki/projects/TOPMed/Freeze.10/Lost_samples/2.indep_study_check/copd_unknown_83_merge_pruned --extract bed1 ",race,"_plink.bed --make-bed --out 83samples_",race))
  bim <- fread(paste0("83samples_",race,".bim"),head=F,data.table=F)
  bim$V2 <- with(bim,paste(V1,V4,sep=":"))
  write.table(bim,paste0("83samples_",race,".bim"),row.names=F,col.names=F,quote=F,sep="\t")
  
  ### Merge
  system(paste0("/udd/redpr/bin/plink19 --bfile COPDGene_",race," --bmerge 83samples_",race," --make-bed --out OVL_merge_",race))
  
  ### King
  system(paste0("/udd/redpr/bin/king -b OVL_merge_",race,".bed --duplicate --prefix king_",race))

  ## Remove tmp files
  system(paste0("rm -f GWAS_",race,"_hg38.*"))
  system(paste0("rm -f COPDGene_",race,".*"))
  system(paste0("rm -f 83samples_",race,".*"))
  system(paste0("rm -f OVL_merge_",race,".*"))
  system(paste0("rm -f ",race,"_plink.bed"))
}
```

#### Result
```{r echo=F}
show.res <- function(race=c("AA","NHW")){
  cat(paste0("#################### ",race," ####################\n\n"))
  con <- read.table(paste0("king_",race,".con"),head=T)
  con <- con[,-c(1,3)]
  within.dup.idx <- which(apply(con[,1:2],1,function(x) length(grep("NWD",x))%in%c(0,2)))
  cat("##### Duplicates within studies #####\n")
  
  if(length(within.dup.idx)>0){
    cat(paste0(race," has ",length(within.dup.idx)," duplicates within studies."))
    print(kable(con[within.dup.idx,]))
    con.btw <- con[-within.dup.idx,]
    cat("\n\n")
  } else {
    cat(paste0(race," has no duplicates within studies.\n"))
    con.btw <- con
  }


  ## ID mapping files
  cat("\n##### Duplicates between studies #####\n")
  if(nrow(con.btw)>0){
    map <- read.csv("/proj/regeps/regep00/studies/COPDGene/analyses/rewki/TOPMed/Freeze.8/0.data/1.WGS/2.QC/Cho_NWD_freeze8_20200214.csv",head=T)
    cat(paste0(race," has ",nrow(con.btw)," duplicates between 83 samples and COPDGene genotyping data"))
    new.con.btw <- con.btw %>%
      left_join(x=.,y=map,by = c("ID2" = "sid")) %>%
      select(ID1,ID2,ID2.NWD=nwd_id,N,N_IBS0,N_IBS1,N_IBS2,Concord,HomConc,HetConc) %>%
      mutate(ID_match= ID1==ID2.NWD)
    print(kable(new.con.btw))
    cat("\n\n")
    return(new.con.btw)
    
  } else {
    cat(paste0(race," has no duplicates between studies.\n"))
    return(NULL)
  }
}

con.btw <- do.call(rbind,lapply(c("AA","NHW"),show.res))
cat(paste0("#################### ID mismatch ####################\n\n"))
ID.mismatch <- con.btw[!con.btw$ID_match,]
rownames(ID.mismatch) <- c()
print(kable(ID.mismatch))
write.csv(con.btw,"COPDGeneGT_dups.csv",row.names=F,quote=F)


## Summary
losts <- read.csv("/udd/rewki/projects/TOPMed/Freeze.10/Lost_samples/New_study_83samples.csv",head=T)
aa <- which(losts$NWD_ID%in%con.btw$ID1)
losts$new.study[aa[-grep("GECOPD",losts$new.study[aa])]] <- paste0(losts$new.study[aa[-grep("GECOPD",losts$new.study[aa])]],";GECOPD") 
losts$new.study <- gsub("^;","",losts$new.study)
losts$source[aa] <- paste0(losts$source[aa],";COPDGene_GT")
losts$source <- gsub("^;","",losts$source)
write.csv(losts,"losts.New_study_83samples.csv",row.names=F,quote=F)
```

### COPDGene exome chip data
- /proj/regeps/regep00/studies/COPDGene/data/exome/CleanExomeChip/data/canonical/[AA,NHW]exomeQCed.[bed,bim,fam]
- Build: hg19

### EOCOPD, ICGN, TCGS
- /proj/regeps/regep00/studies/EOCOPD/data/dna/exome_chip/BWH_Silverman_Exome6k/data/canonical/
    - exome6kSubAndMarkCleanV03.bed
    - exome6kSubAndMarkCleanV03_renamed.bim
    - exome6kSubAndMarkCleanV03.fam
- Build: hg19   
    
#### Result
```{r echo=F}
con <- read.table("king.con",head=T)
lost_sp <- fread(paste0("/udd/rewki/projects/TOPMed/Freeze.10/Lost_samples/2.indep_study_check/copd_unknown_83_merge_pruned.fam"),head=F,data.table=F)$V2

concord.losts <- con[con$ID1%in%lost_sp | con$ID2%in%lost_sp,]
cat(paste0("Among ",nrow(con)," pairs of duplicates, ",nrow(concord.losts)," pairs contain *Lost* samples.\n"))
```



## Summary
```{r echo=F}
losts <- read.csv("/udd/rewki/projects/TOPMed/Freeze.10/Lost_samples/New_study_83samples.csv",head=T)
new.assigned <- losts[losts$new.study!="",]
cat(paste0("Among 83 lost samples, ",nrow(new.assigned)," samples were newly assigned to COPD studies.\n"))
rownames(new.assigned) <- c()
print(kable(new.assigned))
```












```{r eval=F,echo=F}
## JZ's concordance..
library(data.table)
studies <- c("axiom1","exome","GECOPD","CAMP","CRA","ECLPSE")
cons <- do.call(rbind,lapply(studies,function(study) return(fread(paste0("/proj/regeps/regep00/studies/TopMed/data/dna/whole_genome/TopMed/data/freezes/freeze.10.cdnm/tmp/IRCALL_",study,"_duplicate.con"),head=T,data.table=F))))
```

